src/tool_operate.c

```diff
@@ -656,16 +656,106 @@ static void single_transfer_cleanup(struct OperationConfig *config)
   }
 }
 
+static char *ipfs_gateway(void)
+{
+  char *gateway;
+  char *ipfs_path;
+  char *gateway_file_path;
+  FILE *gateway_file;
+
+  gateway = curlx_getenv("IPFS_GATEWAY");
+
+  /* Gateway is found from environment variable. */
+  if(gateway) {
+    bool add_slash = (gateway[strlen(gateway) - 1] == '/') ? FALSE : TRUE;
+    gateway = aprintf("%s%s", gateway, (add_slash) ? "/" : "");
+    return gateway;
+  }
+
+  /* Try to find the gateway in the IPFS data folder. */
+
+  ipfs_path = curlx_getenv("IPFS_PATH");
+
+  if(!ipfs_path) {
+    /* Empty path, fallback to "~/.ipfs", as that's the default location. */
+    ipfs_path = aprintf("%s/.ipfs/", curlx_getenv("HOME"));
+  }
+
+  gateway_file_path = aprintf("%sgateway", ipfs_path);
+  gateway_file = fopen(gateway_file_path, FOPEN_READTEXT);
+
+  if(gateway_file) {
+    char gateway_buffer[255];
+    if(!fgets(gateway_buffer, sizeof(gateway_buffer) - 1, gateway_file)) {
+      fclose(gateway_file);
+      return NULL;
+    }
+    fclose(gateway_file);
+
+    /* Replace the last char with \0 */
+    gateway_buffer[sizeof(gateway_buffer) - 1] = 0;
+
+    /* Replace first occurence of end of line with \0 */
+    gateway_buffer[strcspn(gateway_buffer, "\r")] = 0;
+    gateway_buffer[strcspn(gateway_buffer, "\n")] = 0;
+
+    if(strlen(gateway_buffer) > 1) {
+      bool add_slash = (gateway_buffer[strlen(gateway_buffer) - 1] == '/')
+        ? FALSE
+        : TRUE;
+
+      gateway = aprintf("%s%s", gateway_buffer, (add_slash) ? "/" : "");
+      return gateway;
+    }
+  }
+
+  return NULL;
+}
+
+/*
+ * Rewrite ipfs://<cid> and ipns://<cid> to be a normal
+ * URL that can be handled by an IPFS gateway.
+ */
+static void ipfs_url_rewrite(char **url)
+{
+  char *inner = *url;
+  char *gateway;
+  const char *protocol;
+  char *everything_after_scheme;
+  char *urlbuffer;
+
+  /* 7 is "ipfs://" and "ipns://", it needs o be at least 7 characters*/
+  if(strlen(inner) < 7) {
+    return;
+  }
+
+  /* Get IPFS gateway from environment variable */
+  gateway = ipfs_gateway();
+  protocol = (curl_strnequal(inner, "ipfs", 4)) ? "ipfs" : "ipns";
+
+  everything_after_scheme = &inner[7];
+  urlbuffer = aprintf("%s%s/%s", gateway, protocol, everything_after_scheme);
+
+  *url = urlbuffer; /* use our new URL instead! */
+}
+
 /*
  * Return the proto bit for the scheme used in the given URL
  */
-static long url_proto(char *url)
+static long url_proto(char **url, struct OperationConfig *config)
 {
   CURLU *uh = curl_url();
   long proto = 0;
+
+  /* Before we do anything, see if we need to reqrite the URL.*/
+  if(curl_strnequal(*url, "ipfs", 4) || curl_strnequal(*url, "ipns", 4)) {
+    ipfs_url_rewrite(url);
+    config->followlocation = TRUE;
```

Comment:
Why is IPFS so magic that it needs followlocation explicitly set? What if the user wants to check the first response only?

Reply:
@lidel, could you help a bit on this one? I roughly know why it redirects and how to make it work, but not the details and why it would be required for curl to follow these by default.

Reply:
@bagder there are two types of public gateways that can be used here: 
- [path gateway](https://docs.ipfs.tech/how-to/address-ipfs-on-web/#path-gateway) 
  - `GET https://ipfs.io/ipfs/<cid>/path/to/file.jpg` returns the data
- [subdomain gateway](https://docs.ipfs.tech/how-to/address-ipfs-on-web/#subdomain-gateway)
  - `GET https://dweb.link/ipfs/<cid>/path/to/file.jpg` returns HTTP 301 to a subdomain  `https://<cid>.ipfs.dweb.link/path/to/file.jpg`
    - This is why we suggest following redirects by default: it allows a subdomain gateway to be used as a drop-in replacement for path one, making client implementations less complex.
  

In my mind, following the redirect here can be seen as an UX feature: the curl user should be able to use any of the gateways from  [public gateway list](https://ipfs.github.io/public-gateway-checker/), no matter what the "Origin" column shows (it has no meaning outside web browsers).

If we don't redirect by default, then it is not the end of the world, but every tutorial and document will have to use `curl -L ipfs://<cid>` to avoid issues with subdomain gateways.

Reply:
Thank you, this is a good mental model to follow.  Our examples won't have `-L`.

We can't do much about public gateway snooping on users (yet), but when it comes to tampering (MITM), the need for trusting gateway can be eliminated by requesting verifiable response types:   (https://docs.ipfs.tech/reference/http/gateway/#trusted-vs-trustless)

### Verifiable retrieval with curl

In case of `curl`, trustless retrieval of a directory is possible by requesting CID as a CAR stream and piping it into a tool that verifies content hashes.  Does not need to be a full IPFS node, could be CLI util like  [`ipfs-car`](https://www.npmjs.com/package/ipfs-car):

```console
$ curl "https://ipfs.io/ipfs/bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu?format=car" | ipfs-car
$ cd bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu
```